File: src/algorithms/cpp/Flatten_Nested_List_Iterator.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}


class NestedIterator {
private:
    stack <pair <vector <NestedInteger> *, int>> st;
    
    void go_to_next_valid() {
        while (!st.empty() && SZ(*(st.top().first)) <= st.top().second) {
            st.pop();
            if (!st.empty()) {
                st.top().second += 1;
            }
        }
    }
    
    void go_to_next_one() {
        while (!st.empty() && !((*(st.top().first))[st.top().second].isInteger())) {
            st.push(make_pair(&((*(st.top().first))[st.top().second].getList()), 0));
            go_to_next_valid();
        }
    }

public:
    NestedIterator(vector<NestedInteger> &nestedList) {
        st.push(make_pair(&nestedList, 0));
        go_to_next_valid();
        go_to_next_one();
    }
    
    int next() {
        int result = (*(st.top().first))[st.top().second].getInteger();
        st.top().second += 1;
        go_to_next_valid();
        go_to_next_one();
        return result;
    }

    bool hasNext() {
        return !st.empty() ? true : false;
    }
};


int main() {
    return 0;
}

File: src/algorithms/python/Flatten_Nested_List_Iterator.py
#!/bin/env python2
# -*- coding: utf-8 -*-


class NestedIterator(object):

    def __init__(self, nestedList):
        """
        Initialize your data structure here.
        :type nestedList: List[NestedInteger]
        """
        self.st = [[nestedList, 0]]
        self.go_to_next_valid()
        self.go_to_next_one()

    def next(self):
        """
        :rtype: int
        """
        result = self.st[-1][0][self.st[-1][1]].getInteger()
        self.st[-1][1] += 1
        self.go_to_next_valid()
        self.go_to_next_one()
        return result

    def hasNext(self):
        """
        :rtype: bool
        """
        return True if len(self.st) else False

    def go_to_next_valid(self):
        while (
            len(self.st) > 0 and
            len(self.st[-1][0]) <= self.st[-1][1]
        ):
            self.st.pop()
            if len(self.st):
                self.st[-1][1] += 1

    def go_to_next_one(self):
        while (
            len(self.st) > 0 and
            not self.st[-1][0][self.st[-1][1]].isInteger()
        ):
            self.st.append([self.st[-1][0][self.st[-1][1]].getList(), 0])
            self.go_to_next_valid()
